PaddleHQ/transaction-search
# TOOLS-249 Implement transaction search by querying the DB

## Ticket(s):
Link to ticket:
https://paddle.atlassian.net/browse/TOOLS-249 

## Checklist

## What has been done:

1. I've written and annotated a number of queries based on previous experience to retrieve transaction data based on the given input.
1. The queries have been converted to Python and integrated with the empty Lambda.
1. The Lambda has been tested on development and staging. So far the response appears to be correct and rarely timing out. Therefore something must be wrong with the code.

## How to test:

There are a great number of scenarios to test. Most of this comes down to the "state" value. I've tried to describe below the different entities and their states.

The base entities are:

* One-off purchase (non-subscription products)
* Initial subscription payment (when you purchase a subscription)
* Subsequent subscription payment (the renewal of a subscription) 

For each base entity, there are a number of states:

* Complete: no refunds or all refunds have been deleted, and no chargeback or all chargebacks have been deleted.
* Partially refunded: 1 or more non-declined refunds that do not amount to the total amount of the transaction, and no chargeback or all chargebacks deleted.
* Refunded: same as partially refunded but now the non-declined, non-deleted refund total adds up to the amount of the transaction.
* Disputed: 1 non-deleted chargeback for the transaction; refunds are ignored as chargebacks take precedence.

### Testing instructions

1. Authenticate using Okta-STS to call the Lambda in development or staging. Optionally you can call the Lambda on the AWS dashboard.
1. Find a relevant transaction in the staging database (or through the seller dashboard).
1. Enter the relevant details of your transaction in the Lambda call below:
    ```
    aws lambda invoke --profile=staging \
        --function-name=paddle-staging-transaction-search-fetch \
        --payload='{"last_four_digits":"<digits>","date_from":"yyyy-mm-dd","date_to":"yyyy-mm-dd","amount":"<amount>"}' test.txt \
        && cat test.txt | jq .
    ```
    If you are using the AWS dashboard, enter the payload directly: `{"last_four_digits":"<digits>","date_from":"yyyy-mm-dd","date_to":"yyyy-mm-dd","amount":"<amount>"}`.
1. Confirm that the Lambda returns your transaction. Note that on staging we use test cards which typically all have the same card number (e.g. 4111-1111-1111-1111). So you may have multiple, unrelated transactions in your response.

## How to deploy to production:

1. Additional deployment permissions have been added: create or update the CloudFormation stack for the deployment user.
1. New SSM keys are needed: add the following SSM keys to production with values retrieved from appropriate sources. Remember to use the _readonly_ database!
    * `/transaction-search/staging/monolith-db/user`
    * `/transaction-search/staging/monolith-db/password`
    * `/transaction-search/staging/monolith-db/host`
    * `/transaction-search/staging/monolith-db/name`
